Support to set a specify hostName for a CloneVM

matthewei opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

Sometimes we want to set hostname for CloneVM just as the VM which can be set hostname.

we can realize this function based on the following

```yaml
kind: VirtualMachineClone
apiVersion: "clone.kubevirt.io/v1alpha1"
metadata:
  name: testclone
spec:
  # target definitions hostname
  hostname: vmclone
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #10742

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add the possibility to specify vm target hostname during clone
```

matthewei commented:
I forget to add e2e test and unit-test. I will add it soon.

alaypatel07 commented:
/cc for api change review

alaypatel07 commented on api/openapi-spec/swagger.json:
```diff
@@ -21420,6 +21420,10 @@
       "x-kubernetes-list-type": "atomic"
      },
+     "hostname": {
+      "description": "Specifies the hostname of the clone vmi If not specified, the hostname will be set to the name of the source vmi.",
```
nit: what will happen if the source VMI hostname is empty? is it worth document that empty field will be omitted?

matthewei replied:
yes, if we don't set hostname and the vm will use the original name.

alaypatel07 commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmclone-admitter.go:
```diff
@@ -226,6 +232,21 @@ func validateSource(client kubecli.KubevirtClient, vmClone *clonev1alpha1.Virtua
 
+func validateCloneHostNameNotConformingToDNSLabelRules(field *k8sfield.Path, cloneHostname string) (causes []metav1.StatusCause) {
+	if cloneHostname != "" {
+		errors := validation.IsDNS1123Label(cloneHostname)
```
nit: you might be able to create a new function `func validateHostNameNotConfirmingToDNSLabelRules(field *k8sfield.Path, hostname string)` and use the same to be called here and https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go#L1522-L1529 here.

This will make sure that vmClone and VMI both dont re-use the same logic and test cases avoiding redundency

matthewei replied:
OK, Thank you very much. Follow your advice.

alaypatel07 reviewed (changes requested):
In terms of API, the changes look good, added couple of nits. Pending tests, this would be a good add to vmclone, thanks @matthewei

matthewei commented:
/needs-ok-to-test

matthewei commented:
Hi all , I have finished  all test cases (function test and e2s test). Please help me review it. Many thanks!

matthewei commented:
@alaypatel07 Could you help me review this PR code?

matthewei commented:
Is there anyone who can move to next step?

alaypatel07 commented on pkg/virt-api/webhooks/validating-webhook/admitters/validate-k8s-utils.go:
```diff
@@ -53,6 +55,30 @@ var nodeFieldSelectorValidators = map[string]func(string, bool) []string{
+
+	//convert errorList to []metav1.StatusCause
+	for _, validationErr := range errorList {
+		causes = append(causes, metav1.StatusCause{
```
can you please move the function `appendNewStatusCauseForHostNameNotConformingToDNSLabelRules` to this utils file and use the same format of error message?  we should not change the error message for existing stable APIs

alaypatel07 reviewed (changes requested):
added couple of nits apart from that this looks almost ready

alaypatel07 commented:
@xpivarc can we get a ok-to-test on the PR, it seems ready for CI

xpivarc commented:
/retest required

matthewei commented:
> /retest required

Hi @xpivarc , I think we neet `/ok-to-test`

xpivarc commented:
/test all

matthewei commented:
/test pull-kubevirt-build
/test pull-kubevirt-check-tests-for-flakes
/test pull-kubevirt-e2e-k8s-1.29-sig-storage

matthewei commented:
> @matthewei: Cannot trigger testing until a trusted user reviews the PR and leaves an `/ok-to-test` message.

@xpivarc There are some errors for this PR and I have no access to do it. could you help me to do it?

xpivarc commented:
> > @matthewei: Cannot trigger testing until a trusted user reviews the PR and leaves an `/ok-to-test` message.
> 
> @xpivarc There are some errors for this PR and I have no access to do it. could you help me to do it?

Hi @matthewei , the failures seem relevant but I wouldn't say your PR is causing them. There is no need to retest.

alaypatel07 commented:
@matthewei  Failure logs are found [here](https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/10834/pull-kubevirt-check-tests-for-flakes/1745799290531352576/artifacts/k8s-reporter/pods/1_kubevirt_virt-controller-75ff7c575f-scmlf-virt-controller.log)

matthewei commented:
Why not give this PR a `ok-to-test` and I can check the error logs by myself? Is there any concern for this PR?

matthewei commented:
> @matthewei Failure logs are found [here](https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/10834/pull-kubevirt-check-tests-for-flakes/1745799290531352576/artifacts/k8s-reporter/pods/1_kubevirt_virt-controller-75ff7c575f-scmlf-virt-controller.log)

Thanks a lot @alaypatel07 , I feel like it has nothing to do with this PR? I want to debug it further

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/validate-k8s-utils.go:
```diff
@@ -53,6 +56,29 @@ var nodeFieldSelectorValidators = map[string]func(string, bool) []string{
 }
 
+// validateHostNameNotConformingToDNSLabelRules is function that validate spec.hostName and checks if given affinities are valid
+func validateHostNameNotConformingToDNSLabelRules(field *field.Path, hostName string) (causes []metav1.StatusCause) {
```
In order to apply "utils" I'd suggest to make this function more generic.
```suggestion
func validateFieldValueNotConformingToDNSLabelRules(field *field.Path, value string) (causes []metav1.StatusCause) {
```

And of course to adjust `appendNewStatusCauseForHostNameNotConformingToDNSLabelRules` as well.
Also, I'd suggest to adjust the function to receive the exact field instead of the parent; in this way the
`appendNewStatusCauseForHostNameNotConformingToDNSLabelRules` function does not need to to `field.Child("hostname").String()` but `field.String()` is enough.

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmclone-admitter_test.go:
```diff
@@ -378,6 +378,7 @@ var _ = Describe("Validating VirtualMachineClone Admitter", func() {
 			Entry("negation in the beginning", "!mykey/something"),
 		)
 	})
+
```
please revert

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/validate-k8s-utils.go:
```diff
@@ -28,6 +28,9 @@ package admitters
 	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/validation"
```
nit: move near `k8s.io/apimachinery` section

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:
```diff
@@ -156,7 +156,7 @@ func ValidateVirtualMachineInstanceSpec(field *k8sfield.Path, spec *v1.VirtualMa
 	}
 
-	causes = append(causes, validateHostNameNotConformingToDNSLabelRules(field, spec)...)
+	causes = append(causes, validateHostNameNotConformingToDNSLabelRules(field, spec.Hostname)...)
```
With the current implementation, the resulting cause will be wrong.
The `field.Child("hostname")` in `appendNewStatusCauseForHostNameNotConformingToDNSLabelRules` will fail.
As I already suggested,  this change looks good to me, it is the function that should be adjusted.

fossedihelm commented on tests/clone_test.go:
```diff
@@ -146,6 +146,9 @@ var _ = Describe("[Serial]VirtualMachineClone Tests", Serial, func() {
 
 		vmClone.Spec.Source = cloneSourceRef
 		vmClone.Spec.Target = cloneTargetRef
+		if targetHostname != "" {
```
I don't think we need this if statement

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmclone-admitter_test.go:
```diff
@@ -405,6 +406,18 @@ var _ = Describe("Validating VirtualMachineClone Admitter", func() {
+	Context("Set a specify hostName for a CloneVM ", func() {
+		It("should accept valid hostname", func() {
+			vmClone.Spec.Hostname = "test"
+			admitter.admitAndExpect(vmClone, true)
```
maybe it is better to also check the causes

fossedihelm reviewed (commented):
Hey @matthewei! Thanks for the PR!
First of all, can  I ask you to squash the 2 commits? The first one is totally out of the scope of this PR.
I also had a couple of comments below.
Also, can we add a unit test in the `virt-controller/watch/clone/clone_test.go` file?
There is a specific section for the generated target vm.
Thank you!

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/validate-k8s-utils.go:
```diff
@@ -53,6 +55,29 @@ var nodeFieldSelectorValidators = map[string]func(string, bool) []string{
+	return causes
+}
+
+func appendNewStatusCauseForHostNameNotConformingToDNSLabelRules(field *field.Path, causes []metav1.StatusCause, errors []string) []metav1.StatusCause {
+	return append(causes, metav1.StatusCause{
+		Type: metav1.CauseTypeFieldValueInvalid,
+		Message: fmt.Sprintf("%s does not conform to the kubernetes DNS_LABEL rules : %s",
+			field.String(), strings.Join(errors, ", ")),
+		Field: field.String(),
+	})
+}
```
The name function is still  referring to hostname, while this should be generic too.
Please rename it or drop itand move the content directly inside the caller.
Thanks

fossedihelm reviewed (commented):
Thank you! Some renaming leftover.
Still waiting for a unit test and  the squash of the two commits into one.

fossedihelm commented:
Thanks @matthewei! I still see the first commit that should be dropped from this PR!
Also, we need to cover this new feature with a unit test. Feel free to ask if you need more info.
Thank you

matthewei commented:
> Thanks @matthewei! I still see the first commit that should be dropped from this PR! Also, we need to cover this new feature with a unit test. Feel free to ask if you need more info. Thank you

yes, you are right! I made mistake and  I will remove the first commit!

matthewei commented:
> Thanks @matthewei! I still see the first commit that should be dropped from this PR! Also, we need to cover this new feature with a unit test. Feel free to ask if you need more info. Thank you

Hi, sorry for that. I have dropped the first commit.
I have added the E2E test for this PR. Could you check it again?

matthewei commented:
> /hold The PR looks good, but a unit test is required to get this merged. Also, the first commit should be dropped, because it is not adding anything. Last thing, please, update the release note to something human readable: that will be added in the release note of the next release. Something like: `Add the possibility to specify vm target hostname during clone` should be enough. Thanks!

@fossedihelm many thanks for your patient guidance! I have removed the first commit and added the release note.

fossedihelm commented:
> Hi, sorry for that. I have dropped the first commit. I have added the E2E test for this PR. Could you check it again?

No problem at all! e2e tests and unit tests have a different focus. Unit tests are faster and allow us to test the behavior of different part of the code individually.
With a unit test, I'd expect smth like:
```
Context("hostname", func() {
	It("should copy source hostname if hostname is not provided", func() {
            [...]
	})

	It("should set custom hostname if provided", func() {
            [...]
	})
})
```
to be added under this `Context` https://github.com/kubevirt/kubevirt/blob/ad05e07dbfe6a7d0a648fafbf29d0c4aa51b9b47/pkg/virt-controller/watch/clone/clone_test.go#L670
Thank you!

matthewei commented:
> > Hi, sorry for that. I have dropped the first commit. I have added the E2E test for this PR. Could you check it again?
> 
> No problem at all! e2e tests and unit tests have a different focus. Unit tests are faster and allow us to test the behavior of different part of the code individually. With a unit test, I'd expect smth like:
> 
> ```
> Context("hostname", func() {
> 	It("should copy source hostname if hostname is not provided", func() {
>             [...]
> 	})
> 
> 	It("should set custom hostname if provided", func() {
>             [...]
> 	})
> })
> ```
> 
> to be added under this `Context`
> 
> https://github.com/kubevirt/kubevirt/blob/ad05e07dbfe6a7d0a648fafbf29d0c4aa51b9b47/pkg/virt-controller/watch/clone/clone_test.go#L670
> 
> 
> Thank you!

done

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/validate-k8s-utils.go:
```diff
@@ -53,6 +55,25 @@ var nodeFieldSelectorValidators = map[string]func(string, bool) []string{
 }
 
+// validateFieldValueNotConformingToDNSLabelRules is function that validate value and checks if given affinities are valid
+func validateFieldValueNotConformingToDNSLabelRules(field *field.Path, value, valueField string) (causes []metav1.StatusCause) {
```
Can I ask you why you introduced the `valueField` parameter? IMHO it can create confusion.
Also, in this case the `field` parameter is just a parent `Path`.
Are there any reasons that I am not seeing?
If the problem is the caller param, it is enough to declare the `field` as `k8sfield.NewPath("spec", "hostname")`
Thanks

fossedihelm commented on pkg/virt-controller/watch/clone/clone_test.go:
```diff
@@ -1002,6 +1002,39 @@ var _ = Describe("Clone", func() {
+
+			const manuallySetHostName = "manually-set-hostname"
+
+			BeforeEach(func() {
+				sourceVM.Spec.Template.Spec.Hostname = vmClone.Name
+			})
+
+			expectHostName := func(hostname string) {
+				expectedVM := sourceVM.DeepCopy()
+				expectedVM.Spec.Template.Spec.Hostname = hostname
+				expectVMCreationFromPatches(expectedVM)
+			}
+
+			It("should set with the kubevirt generated hostname if hostname is not provided", func() {
+				addClone(vmClone)
+				expectHostName(vmClone.Name)
+				expectSnapshotContentGet(sourceVM)
+
+				controller.Execute()
+			})
```
The `BeforeEach` instruction makes the test `should set with the kubevirt generated hostname if hostname is not provided` not correct.
In the `BeforeEach` you are providing an hostname to the target vmi.
Please, drop the `BeforeEach`.
Thanks

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:
```diff
@@ -156,7 +156,7 @@ func ValidateVirtualMachineInstanceSpec(field *k8sfield.Path, spec *v1.VirtualMa
 	}
 
-	causes = append(causes, validateHostNameNotConformingToDNSLabelRules(field, spec)...)
+	causes = append(causes, validateFieldValueNotConformingToDNSLabelRules(field, spec.Hostname)...)
```
```suggestion
	causes = append(causes, validateFieldValueNotConformingToDNSLabelRules(field.Child("hostname"), spec.Hostname)...)
```

fossedihelm reviewed (changes requested):
We are almost there! Thanks for the patience!

fossedihelm commented:
https://github.com/kubevirt/kubevirt/compare/9a14290264de8ed5bf720624d574d9ed4bce4c91..8a879b7ad065826be22a65f512349c874ea4b039

The change in the file `pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go`
will generate a failure in the unit test:
```
Expect(causes[0].Field).To(Equal("fake.hostname"))
```
Please re-apply.
Thanks

matthewei commented:
> https://github.com/kubevirt/kubevirt/compare/9a14290264de8ed5bf720624d574d9ed4bce4c91..8a879b7ad065826be22a65f512349c874ea4b039
> 
> The change in the file `pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go` will generate a failure in the unit test:
> 
> ```
> Expect(causes[0].Field).To(Equal("fake.hostname"))
> ```
> 
> Please re-apply. Thanks

I `make test` in my env and I can pass all the unit-test.

fossedihelm commented:
> ```
> Expect(causes[0].Field).To(Equal("fake.hostname"))
> ```

You are right! Sorry about that! My bad!
/lgtm
/unhold

matthewei commented:
> > ```
> > Expect(causes[0].Field).To(Equal("fake.hostname"))
> > ```
> 
> You are right! Sorry about that! My bad! Thank you! /lgtm /unhold

Never mind! Thank you for your patient guidance!

matthewei commented:
/retest-required

fossedihelm commented:
/retest

fossedihelm commented:
Sure! Thank you!
/lgtm
/retest

matthewei commented:
> @matthewei I think something went wrong with the rebase. Please check with `make && make generate`. Thank you

HI sir, Happy Chinese New Year! Many thanks. I can execute `make && make generate`.

fabiand commented:
Happ new year! :fireworks:

On Friday, February 23, 2024, matthewei ***@***.***> wrote:
> @matthewei I think something went wrong with the rebase. Please check
with make && make generate. Thank you
>
> HI sir, Happy Chinese New Year! Many thanks. I can execute make && make
generate.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you are subscribed to this thread.<
https://ci4.googleusercontent.com/proxy/JX5BoDm1X-BawZCMqt0mOEK__bzHx28OABZjfnfNXm7ZPmsThx4xSKToYYCTcaICDck0gyWnFt3eRVFjbxd8-KFd4s5wAH1ApyWov4LJBp6u659_gdhZOS3YynjmkM_F1RS3L6hqk1qeWpYGoj1Y2MtszHZpyocZvTQvanowEOXTg6Qg-j3A4kiwDYUReb2sR_Z7dxavbZaO0fqV8_vemibTArGtb1Vdp8cXJrHwHlzJPCgvBA=s0-d-e1-ft#https://github.com/notifications/beacon/AAAUMAFEFTFJUJ4V2HDEB3LYU7TP7A5CNFSM6AAAAABAHOT2OSWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTTU3RK5C.gif>Message
ID: ***@***.***>

fossedihelm reviewed (commented):
Hi @matthewei happy new year! 
As I suspected, the rebase created some conflicts. I checked the code and the reason is the refactored clone test merged a while ago.
Below, some comments on how to solve those, but feel free to use your own style.
Thank you!

fossedihelm commented:
> Hi @matthewei happy new year! As I suspected, the rebase created some conflicts. I checked the code and the reason is the refactored clone test merged a while ago. Below, some comments on how to solve those, but feel free to use your own style. Thank you!

Generally speaking, before any push run `make && make generate` and add the (optional generated) changes to your commit, so that the CI does not fail :)

fossedihelm commented:
/lgtm cancel
Waiting for fixing the failures

fossedihelm commented:
/test all

matthewei commented:
I would like to create a new PR for it.

matthewei commented:
@fossedihelm Sorry bother you again. Please help me review again. Many thanks!

fossedihelm commented on pkg/virt-controller/watch/clone/vm-target.go:
```diff
@@ -197,3 +200,13 @@ func generateFirmwareUUIDPatches(firmware *k6tv1.Firmware) (patches []string) {
 }
+
+func generateHostnamePatches(cloneHostname string) (patches []string) {
+	const hostnamePatchPattern = `{"op": "replace", "path": "/spec/template/spec/hostname", "value": "%s"}`
+
+	if cloneHostname == "" {
+		return nil
+	}
+	patch := fmt.Sprintf(hostnamePatchPattern, cloneHostname)
+	return []string{patch}
```
Would you mind to use the new `PatchSet` https://github.com/kubevirt/kubevirt/pull/11599 ?

fossedihelm reviewed (commented):
Thank you @matthewei! Sorry for not handling this PR. It must have slipped off my todo list.

fossedihelm commented:
/hold
Please run `make && make generate` and commit the changes. Thanks

fossedihelm commented:
/test pull-kubevirt-build pull-kubevirt-generate pull-kubevirt-unit-test pull-kubevirt-verify-go-mod  pull-kubevirt-manifests
/unhold

fossedihelm commented:
@matthewei Looks like there are some other changes to be committed.
Can you try again `make && make generate`? Thanks

matthewei commented:
Please hold this PR.

fossedihelm commented:
/hold

matthewei commented:
@fossedihelm Good moring my friend!. I have `make` and `make generate` in my local env and make a full test for this PR. Please help me review it.

matthewei commented:
could you give this PR a "/test" label?

fossedihelm commented:
/test pull-kubevirt-build pull-kubevirt-generate pull-kubevirt-unit-test pull-kubevirt-verify-go-mod pull-kubevirt-manifests

fossedihelm commented:
/ok-to-test

matthewei commented:
Wow, Why are the checks always pending? Is there anyone who can have a look?

fossedihelm commented:
/ok-to-test

fossedihelm commented:
/test all

fossedihelm commented:
/unhold

matthewei commented:
/retest-required

matthewei commented:
/ok-to-test

fossedihelm commented on tests/clone_test.go:
```diff
@@ -412,6 +414,30 @@ var _ = Describe("[Serial]VirtualMachineClone Tests", Serial, func() {
+				createCloneAndWaitForFinish(vmClone)
+
+				By(fmt.Sprintf("Getting the target VM %s", targetVMName))
+				targetVM, err = virtClient.VirtualMachine(sourceVM.Namespace).Get(context.Background(), targetVMName, &v1.GetOptions{})
```
VirtualMachine doesn't use pointers anymore:
```suggestion
				targetVM, err = virtClient.VirtualMachine(sourceVM.Namespace).Get(context.Background(), targetVMName, v1.GetOptions{})
```

fossedihelm reviewed (commented):
/ok-to-test

fossedihelm commented:
@brianmcarey Looks like `/ok-to-test` does not work here. Can you have a look? Thanks
/test all

matthewei commented:
@brianmcareyCould  you have a look? Thanks

fossedihelm commented:
@matthewei I asked around about the `ok-to-test` and it was disabled, this is the reason why it does not work.
/retest

matthewei commented:
@fossedihelm Moring! I have checked the error's log and I find it has nothing with this PR at all, Could you spare a minute to have a look? Many thanks!

fossedihelm commented:
> @fossedihelm Moring! I have checked the error's log and I find it has nothing with this PR at all, Could you spare a minute to have a look? Many thanks!

@matthewei Sure! Definitely not this PR related failure. Maybe just a flake test we need to investigate.
/retest-required

matthewei commented:
/retest

matthewei commented:
Could you give me the access to set `retest`? I rememered that somethime I can set `retest`.

fossedihelm commented:
/test pull-kubevirt-e2e-k8s-1.29-sig-compute
@matthewei As @brianmcarey mentioned the `/ok-to-test` has been set to be ignored for security reasons, so now only org members can run `/test`.
This is quite recent.
Feel free to ask me for retest. 
@iholder101 would you mind having a look?
Thank you

iholder101 commented on tests/clone_test.go:
```diff
@@ -136,7 +136,7 @@ var _ = Describe("[Serial]VirtualMachineClone Tests", Serial, func() {
 	}
 
-	generateCloneFromVMWithParams := func(sourceVM *virtv1.VirtualMachine, targetVMName string) *clonev1alpha1.VirtualMachineClone {
+	generateCloneFromVMWithParams := func(sourceVM *virtv1.VirtualMachine, targetVMName, hostName string) *clonev1alpha1.VirtualMachineClone {
```
Please avoid adding the hostname as a parameter here and define it in the specific test that's relevant

iholder101 commented on tests/clone_test.go:
```diff
@@ -412,6 +414,30 @@ var _ = Describe("[Serial]VirtualMachineClone Tests", Serial, func() {
+				By("Making sure target is runnable")
+				targetVM = expectVMRunnable(targetVM)
+
+				By("Making sure hostname is applied to target VM")
+				Expect(targetVM.Spec.Template.Spec.Hostname).ToNot(BeNil())
```
This field is a string, hence cannot be nil, so this is redundant IMO

iholder101 reviewed (commented):
Thansk @matthewei!
See comments below

matthewei commented:
> @matthewei The failure is strange. Can you rebase on main?

yes, I will work on it.

matthewei reviewed (commented):
Plese help me to review this PR

matthewei commented:
@fossedihelm Hi, please help me review this PR. Many thanks!

matthewei commented:
/retest

iholder101 commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmclone-admitter.go:
```diff
@@ -108,6 +110,10 @@ func (admitter *VirtualMachineCloneAdmitter) Admit(ar *admissionv1.AdmissionRevi
 		causes = append(causes, newCauses...)
 	}
 
+	if newCauses := validateCloneHostNameNotConformingToDNSLabelRules(vmClone); newCauses != nil {
```
In the future we may add more general validations to hostname
```suggestion
	if newCauses := validateCloneHostName(vmClone); newCauses != nil {
```

matthewei replied:
done

iholder101 reviewed (commented):
Thanks a lot @matthewei!
Almost there! :) left some small comments.

Great job and thanks for your patience!

matthewei commented on staging/src/kubevirt.io/api/clone/v1alpha1/types.go:
```diff
@@ -65,7 +65,9 @@ type VirtualMachineCloneSpec struct {
 	// +optional
 	Target *corev1.TypedLocalObjectReference `json:"target,omitempty"`
-
+	// Specifies the hostname of the clone vmi.
```
done

iholder101 commented:
Thanks a lot @matthewei!
/test all

iholder101 commented:
@matthewei There are lane failures. For example, the unit tests lane says:
```
pkg/virt-api/webhooks/validating-webhook/admitters/vmclone-admitter_test.go:409:28: cannot use "test" (untyped string constant) as *string value in assignment
pkg/virt-api/webhooks/validating-webhook/admitters/vmclone-admitter_test.go:414:28: cannot use "test+bad" (untyped string constant) as *string value in assignment
```

Can you please to build and test locally? To do so:
* Run `make`
* Run `make test`
* Optional: change `Describe` to `FDescribe` here and run `make cluster-up && make cluster-sync && make functest` https://github.com/kubevirt/kubevirt/blob/ddc9d60c79d7ec7fdef21c4299300d8154441398/tests/clone_test.go#L50

matthewei commented:
> @matthewei There are lane failures. For example, the unit tests lane says:
> 
> ```
> pkg/virt-api/webhooks/validating-webhook/admitters/vmclone-admitter_test.go:409:28: cannot use "test" (untyped string constant) as *string value in assignment
> pkg/virt-api/webhooks/validating-webhook/admitters/vmclone-admitter_test.go:414:28: cannot use "test+bad" (untyped string constant) as *string value in assignment
> ```
> 
> Can you please to build and test locally? To do so:
> 
> * Run `make`
> * Run `make test`
> * Optional: change `Describe` to `FDescribe` here and run `make cluster-up && make cluster-sync && make functest` https://github.com/kubevirt/kubevirt/blob/ddc9d60c79d7ec7fdef21c4299300d8154441398/tests/clone_test.go#L50

OK, will do it

matthewei commented:
> @matthewei There are lane failures. For example, the unit tests lane says:
> 
> ```
> pkg/virt-api/webhooks/validating-webhook/admitters/vmclone-admitter_test.go:409:28: cannot use "test" (untyped string constant) as *string value in assignment
> pkg/virt-api/webhooks/validating-webhook/admitters/vmclone-admitter_test.go:414:28: cannot use "test+bad" (untyped string constant) as *string value in assignment
> ```
> 
> Can you please to build and test locally? To do so:
> 
> * Run `make`
> * Run `make test`
> * Optional: change `Describe` to `FDescribe` here and run `make cluster-up && make cluster-sync && make functest` https://github.com/kubevirt/kubevirt/blob/ddc9d60c79d7ec7fdef21c4299300d8154441398/tests/clone_test.go#L50

Hi， I got the following errors,Could you give me a hand?
![image](https://github.com/kubevirt/kubevirt/assets/86186853/bb55f236-0a70-430b-afa7-a29ea21fd553)

iholder101 commented:
@matthewei you probably need to deploy storage to the cluster. Try:
```bash
> export KUBEVIRT_STORAGE=rook-ceph-default
> make cluster-up && make cluster-sync
```

matthewei commented:
> ```shell
> make cluster-up && make cluster-sync
> ```

ok , Come back and ready going on it

iholder101 commented:
> @iholder101 If I want to exec the command `test all`. How can i realize it?

In order to do so you need to be a community member.
Let me help you with that.

/test pull-kubevirt-build
/test pull-kubevirt-unit-test
/test pull-kubevirt-e2e-k8s-1.30-sig-compute
/test pull-kubevirt-e2e-k8s-1.30-sig-compute

I also suggest to try and building, running unit tests, and preferably some functional tests before triggering CI.

iholder101 commented:
Are you still interested in pushing this forward? @matthewei
